### PR TITLE
Update Branch Protection to current GitHub API

### LINKED ISSUE #586: Add support for advanced branch protection APIs
It seems like the following APIs exist in github v3 (as beta?) but are not yet reflected in this library.

- https://developer.github.com/v3/repos/branches/#get-pull-request-review-enforcement-of-protected-branch
- https://developer.github.com/v3/repos/branches/#update-pull-request-review-enforcement-of-protected-branch
- https://developer.github.com/v3/repos/branches/#get-admin-enforcement-of-protected-branch
- https://developer.github.com/v3/repos/branches/#get-restrictions-of-protected-branch
- and a few others....

It'd be great to have full support of this API (especially since it allows controlling PR reviews and how many are needed and by who).

### PR DESCRIPTION
**Pull request by s-t-e-v-e-n-k**
The Branch protection API has been radically changed by GitHub, add new
methods to Branch reflecting it. Branch protection methods are now
called on the Branch object itself, rather than awkwardly hanging off the
Repository object and requiring branch names to be passed in.

This adds an over-arching methods to get, edit and remove protection
entirely, as well as fine-grained methods for getting, setting and
dropping certain aspects of branch protection, as well as three new
objects to encompass querying those aspects.

This also destroys Repository.protect_branch(), the endpoint has been
removed. The old-style protection attributes on Branch have also been
removed to force users onto the new API, since they are still sent, but
no longer populated.

Fixes #586

--

I'm sorry for just how large this branch is. I tried to split it up, but everything depends on the new methods in Branch, which pretty much pulls in everything ever due to the tests.

### CODE REVIEW

**Comment by JPWKU** [comment]
@s-t-e-v-e-n-k I am reviewing, but will take a few hours at least.

File: github/Branch.py [hunk 1]
```diff
@@ -67,29 +71,282 @@ def protected(self):
...
+    def remove_admin_enforcement(self):
+        """
+        :calls: `DELETE /repos/:owner/:repo/branches/:branch/protection/enforce_admins <https://developer.github.com/v3/repos/branches>`_
+        """
+        headers, data = self._requester.requestJsonAndCheck(
+            "DELETE",
+            self.protection_url + "/enforce_admins"
+        )
+
+    def get_user_push_restrictions(self):
```
Comment (JPWKU) [inline github/Branch.py]: +1

File: github/Branch.py [hunk 2]
```diff
@@ -67,29 +71,282 @@ def protected(self):
...
+        :rtype: :class:`github.PaginatedList.PaginatedList` of :class:`github.NamedUser.NamedUser`
+        """
+        return github.PaginatedList.PaginatedList(
+            github.NamedUser.NamedUser,
+            self._requester,
+            self.protection_url + "/restrictions/users",
+            None
+        )
+
+    def get_team_push_restrictions(self):
```
Comment (JPWKU) [inline github/Branch.py]: +1

File: github/Branch.py [hunk 3]
```diff
@@ -67,29 +71,282 @@ def protected(self):
...
+        """
+        assert all(isinstance(element, (str, unicode)) or isinstance(element, (str, unicode)) for element in users), users
+
+        headers, data = self._requester.requestJsonAndCheck(
+            "POST",
+            self.protection_url + "/restrictions/users",
+            input=users
+        )
+
+    def edit_team_push_restrictions(self, *teams):
```
Comment (JPWKU) [inline github/Branch.py]: +1

File: github/Branch.py [hunk 4]
```diff
@@ -67,29 +71,282 @@ def protected(self):
...
+            post_parameters["required_status_checks"] = {"strict": strict, "contexts": contexts}
+        else:
+            post_parameters["required_status_checks"] = None
+
+        if enforce_admins is not github.GithubObject.NotSet:
+            post_parameters["enforce_admins"] = enforce_admins
+        else:
+            post_parameters["enforce_admins"] = None
+
+        if dismissal_users is not github.GithubObject.NotSet or dismissal_teams is not github.GithubObject.NotSet or dismiss_stale_reviews is not github.GithubObject.NotSet or require_code_owner_reviews is not github.GithubObject.NotSet:
```
[the remark this answers is not in the document]
Comment (s-t-e-v-e-n-k) [reply github/Branch.py]: "The Protected Branches API now has a setting for requiring a specified number of approving pull request reviews before merging. This feature is currently available for developers to preview."

This requires sending a custom Accept header which the code doesn't do. I do plan to add it, but this branch was already large enough.

File: github/Branch.py [hunk 5]
```diff
@@ -67,29 +71,282 @@ def protected(self):
...
+                post_parameters["required_pull_request_reviews"]["require_code_owner_reviews"] = require_code_owner_reviews
+            if dismissal_users is not github.GithubObject.NotSet:
+                post_parameters["required_pull_request_reviews"]["dismissal_restrictions"] = {"users": dismissal_users}
+            if dismissal_teams is not github.GithubObject.NotSet:
+                if "dismissal_restrictions" not in post_parameters["required_pull_request_reviews"]:
+                    post_parameters["required_pull_request_reviews"]["dismissal_restrictions"] = {}
+                post_parameters["required_pull_request_reviews"]["dismissal_restrictions"]["teams"] = dismissal_teams
+        else:
+            post_parameters["required_pull_request_reviews"] = None
+        if user_push_restrictions is not github.GithubObject.NotSet or team_push_restrictions is not github.GithubObject.NotSet:
```
[the remark this answers is not in the document]
Comment (s-t-e-v-e-n-k) [reply github/Branch.py]: To my mind, the exception in itself is useful enough here -- that code is checked in the tests.

File: github/Branch.py [hunk 6]
```diff
@@ -67,29 +71,282 @@ def protected(self):
...
+
+    def edit_required_pull_request_reviews(self, dismissal_users=github.GithubObject.NotSet, dismissal_teams=github.GithubObject.NotSet, dismiss_stale_reviews=github.GithubObject.NotSet, require_code_owner_reviews=github.GithubObject.NotSet):
+        """
+        :calls: `PATCH /repos/:owner/:repo/branches/:branch/protection/required_pull_request_reviews <https://developer.github.com/v3/repos/branches>`_
+        """
+        assert dismissal_users is github.GithubObject.NotSet or all(isinstance(element, (str, unicode)) or isinstance(element, (str, unicode)) for element in dismissal_users), dismissal_users
+        assert dismissal_teams is github.GithubObject.NotSet or all(isinstance(element, (str, unicode)) or isinstance(element, (str, unicode)) for element in dismissal_teams), dismissal_teams
+        assert dismiss_stale_reviews is github.GithubObject.NotSet or isinstance(dismiss_stale_reviews, bool), dismiss_stale_reviews
+        assert require_code_owner_reviews is github.GithubObject.NotSet or isinstance(require_code_owner_reviews, bool), require_code_owner_reviews
+
```
[the remark this answers is not in the document]
Comment (s-t-e-v-e-n-k) [reply github/Branch.py]: As above, need the custom Accept header for that.

**Comment by JPWKU** [comment]
Circling back to this today..been a busy week.

**Comment by sfdye** [comment]
Finally someone taking this up. Good job @s-t-e-v-e-n-k

**Comment by jiang-wei** [comment]
waiting for the merge for days ...

**Comment by sfdye** [comment]
@JPWKU I really want to get this out. Do you have anything to add for the code review?

**Comment by JPWKU** [comment]
@jiang-wei @sfdye I realize this has been in limbo for awhile.  Just a combination of it being a large PR and having limited bandwidth to go over.  Thanks for bringing it to the top again.  @s-t-e-v-e-n-k Looks like there is a conflict, can you rebase and resolve that to see a test run again?

**Comment by s-t-e-v-e-n-k** [comment]
@JPWKU Done! Let's get this merged! \o/

**Comment by sfdye** [comment]
LGTM. @s-t-e-v-e-n-k Could you please add a summary for this PR, for the release note? I don't wanna miss anything 😄

**Comment by s-t-e-v-e-n-k** [comment]
@sfdye I think the following:

"Add support for the Branch Protection API, which adds 12 new methods onto github.Branch allowing repository owners to set protection on any branch, such as user or team push restrictions, or requiring a number of reviewers before a PR can be merged."

Is that enough, or do I need to go into more detail?

**Comment by sfdye** [comment]
It's fine, I will reformat to release note style bullet points.

**Comment by sfdye** [comment]
@s-t-e-v-e-n-k 

I guess with this PR merged, `Repository.get_protected_branch` should also be removed right?

https://github.com/PyGithub/PyGithub/blob/080d807552718df5179e7998cc245e3bb9ea0262/github/Repository.py#L1314:L1325

**Comment by s-t-e-v-e-n-k** [comment]
@sfdye I had a PR that was marked as stale that deprecated it, but this PR already adds a breaking change, so I'll push up a PR tonight that removes it.

**Comment by sfdye** [comment]
Oh yeah, we can just remove it. I have issued a warning on the release note as well.

**Pull request merged**